### ISSUE TITLE
TOML Config content rearranged on second launch

### ISSUE THREAD

**Issue by EwyBoy** [MinecraftForge/MinecraftForge#7489]
**Minecraft Version:**
1.16.4 - (but been this way since the new TOML config system was released).

**Forge Version:**
35.0.15

**Steps to Reproduce:**
 1. Delete config if you have one.
 2. Launch game to generate a fresh config file.
 3. Look at config file (Looks nice and tidy, in the order you coded it to generate).
 4. Close game.
 5. Relaunch game.
 6. Look at config file (Now its all messed up and things are shifted around).

**Description of issue:**
The TOML config generated by the ForgeConfigSpec ~~reorganizes~~ deorganizes the config file on the launch **after** it is generated.

### Pictures:
**Freshly generated config:**
![Config_first](https://i.imgur.com/h3ny73l.png)

**Config after second launch:**
![Config_second](https://i.imgur.com/LvSr5Vg.png)

### Source code
```java
public class ConfigTest {

    private static final Builder BUILDER = new Builder();

    public static final class CategoryGeneral {

        public final ForgeConfigSpec.ConfigValue<List<? extends String>> test_1;
        public final ForgeConfigSpec.ConfigValue<List<? extends String>> test_2;
        public final ForgeConfigSpec.ConfigValue<List<? extends String>> test_3;
        public final ForgeConfigSpec.ConfigValue<List<? extends String>> test_4;
        public final ForgeConfigSpec.ConfigValue<List<? extends String>> test_5;

        public static List<String> myList = new ArrayList<>();

        static {
            myList.add("Test 1");
            myList.add("Test 2");
            myList.add("Test 3");
        }

        private CategoryGeneral() {

            BUILDER.comment("General mod settings").push("general");

            test_1 = BUILDER.defineList("test_1", Lists.newArrayList(myList), o -> o instanceof String);
            test_2 = BUILDER.defineList("test_2", Lists.newArrayList(myList), o -> o instanceof String);
            test_3 = BUILDER.defineList("test_3", Lists.newArrayList(myList), o -> o instanceof String);
            test_4 = BUILDER.defineList("test_4", Lists.newArrayList(myList), o -> o instanceof String);
            test_5 = BUILDER.defineList("test_5", Lists.newArrayList(myList), o -> o instanceof String);

            BUILDER.pop();
        }
    }

    public static final ForgeConfigSpec CONFIG = BUILDER.build();

}
```
I have traced this bug back as far as the new TOML config system goes, looking at [World Stripper](https://www.curseforge.com/minecraft/mc-mods/world-stripper)'s config file who has been affected by this issue for a long while now.

Here I am allowing the user to store up to 5 "profiles" for world editing in the config. But it confuses people when the profiles gets shifted around and get out of order.

**Freshly generated config:**
![Config_first_ws](https://i.imgur.com/ecAhAmD.png)

**Config after second launch:**
![Config_second_ws](https://i.imgur.com/GAI8xGg.png)

Here you can also see that the categories (pushes / pops) gets rearranged.

@ me for more info on this subject.

**Comment by TheCurle** [comment]
It's possible this may be causing the fml.toml corruption we've been seeing for a while.
This warrants further investigation.

Thanks for the detailed report.

**Comment by EwyBoy** [comment]
Looks like this issue was reported before and "fixed". But for some reason it has been unfixed or not properly fixed.

Earlier reports: #7224 and #5688
Pull request with fix: #7230

**Comment by diesieben07** [comment]
This is a bug in NightConfig. [Here](https://github.com/TheElectronWill/night-config/blob/78d9b1cbfc700bacddab903576693f7b89306a80/toml/src/main/java/com/electronwill/nightconfig/toml/TomlParser.java#L90-L93) a sub-config is read. However it fails to inherit the map creator for this sub config, causing the sub config to use a simple HashMap instead of the configured `LinkedHashMap`.

**Comment by diesieben07** [comment]
https://github.com/TheElectronWill/night-config/issues/85

**Comment by Eric-01** [comment]
Since the linked issue has already been fixed and the related pull request merged, what is now needed to make it work in Forge?

**Comment by EwyBoy** [comment]
Add this before your config is generated fixes the order of config values.
```Java
import com.electronwill.nightconfig.core.Config;

Config.setInsertionOrderPreserved(true);
```

**Comment by diesieben07** [comment]
night-config has not yet released a new version containing the fix, so there is nothing that can be done in FOrge for the time being.

**Comment by sciwhiz12** [comment]
NightConfig has a new version, [v3.6.4](https://github.com/TheElectronWill/night-config/releases/tag/v3.6.4), which contains the fix (by virtue of the release being on the stable-3.x branch, where @diesieben07's fix was merged to).

**Comment by sciwhiz12** [comment]
Fixed in 1.18.x by eb1a9b57c6cc4d5cb4ad3c4305c2a73015d28917 with the update of NightConfig to v3.6.4. Contributors are welcome to backport the update to LTS.
